llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -1854,12 +1855,53 @@ bool MemCpyOptPass::processMemCpy(MemCpyInst *M, BasicBlock::iterator &BBI) {
   return false;
 }
 
+/// Memmove calls with overlapping src/dest buffers that come after a memset may
+/// be removed.
+bool MemCpyOptPass::isMemMoveMemSetDependency(MemMoveInst *M) {
+  MemoryUseOrDef *MemMoveAccess = MSSA->getMemoryAccess(M);
+  if (!MemMoveAccess)
+    return false;
+
+  BatchAAResults BAA(*AA);
+  MemoryAccess *FirstDef = MemMoveAccess->getDefiningAccess();
+  MemoryLocation SourceLoc = MemoryLocation::getForSource(M);
+  MemoryAccess *SourceClobber =
+      MSSA->getWalker()->getClobberingMemoryAccess(FirstDef, SourceLoc, BAA);
+
+  MemSetInst *MS = nullptr;
+  if (auto *Def = dyn_cast<MemoryDef>(SourceClobber))
+    MS = dyn_cast_or_null<MemSetInst>(Def->getMemoryInst());
+
+  if (!MS || MS->getParent() != M->getParent())
```

Comment:
Please add a negative test for `MS->getParent() != M->getParent()`. Why we cannot do this if the memset and memmove are not in the same block?

Reply:
I had limited this to the same block mostly as to avoid unnecessaries queries to the cached AA results, but I think it should be fine as long as there are no clobbering writes in between. Maybe worth understanding if this is the correct way to pursue, before dropping it.